src/_pytest/scope.py

```diff
@@ -0,0 +1,79 @@
+"""
+Scope definition and related utilities.
+
+Those are defined here, instead of in the 'fixtures', module because
+their use is spread across many other pytest modules, and centralizing it in 'fixtures'
+would cause circular references.
+
+Also this makes the module light to import, as it should.
+"""
+from enum import Enum
+from functools import lru_cache
+from typing import Optional
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing_extensions import Literal
+
+    _ScopeName = Literal["session", "package", "module", "class", "function"]
+
+
+class Scope(Enum):
+    """Represents one of the possible fixture scopes in pytest."""
+
+    Session = "session"
+    Package = "package"
+    Module = "module"
+    Class = "class"
+    Function = "function"
+
+    def index(self) -> int:
+        """Return this scope index. Smaller numbers indicate higher scopes (Session = 0)."""
+        return _scope_to_index(self)
+
+    def next(self) -> "Scope":
+        """Return the next scope (from top to bottom)."""
+        return _next_scope(self)
+
+    @classmethod
+    def from_user(
+        cls, scope_name: "_ScopeName", descr: str, where: Optional[str] = None
+    ) -> "Scope":
+        """
+        Given a scope name from the user, return the equivalent Scope enum. Should be used
+        whenever we want to convert a user provided scope name to its enum object.
+
+        If the scope name is invalid, construct a user friendly message and call pytest.fail.
+        """
+        from _pytest.outcomes import fail
+
+        try:
+            return Scope(scope_name)
+        except ValueError:
+            fail(
+                "{} {}got an unexpected scope value '{}'".format(
+                    descr, f"from {where} " if where else "", scope_name
+                ),
+                pytrace=False,
+            )
+
+
+# Ordered list of scopes which can contain many tests (in practice all except Function).
+HIGH_SCOPES = [x for x in Scope if x is not Scope.Function]
+
+
+@lru_cache(maxsize=None)
+def _scope_to_index(scope: Scope) -> int:
+    """Implementation of Scope.index() as a free function so we can cache it."""
+    scopes = list(Scope)
+    return scopes.index(scope)
+
+
+@lru_cache(maxsize=None)
+def _next_scope(scope: Scope) -> Scope:
```

Comment:
part of me wonders if the declaration of next_scope shouldnt just be a plain old mapping

we could pretty much do with a ordered mapping there

Reply:
If you prefer something more explicit, probably we can then use this:

```python
@lru_cache(maxsize=None)
def _next_scope(scope: Scope) -> Scope:
    if scope is Scope.Function:
        raise ValueError("Function is the bottom scope")
    if scope is Scope.Session:
        return Scope.Package
    if scope is Scope.Package:
        return Scope.Module
    if scope is Scope.Module:
        return Scope.Class
    if scope is Scope.Class:
        return Scope.Function
    assert_never(scope)
```

So `assert_never` will fail if we ever add a new scope and forget to update this method.

However TBH I think the current implementation is OK, but don't have strong feelings either.

Reply:
no, what i meant is 

```py
NEXT_SCOPES = {
  Scope.Session: Scope.Package,
  ....
  }
```

Reply:
I know, but I was proposing the above because then we can use `assert_never`.

However we still need a function to be able to raise a proper error when `Scope.Function` is given, that's why I'm not sure having a mapping buys us much:

```
@lru_cache(maxsize=None)
def _next_scope(scope: Scope) -> Scope:
    if scope is Scope.Function:
        raise ValueError("Function is the bottom scope")
    return _NEXT_SCOPES[scope]
```

Reply:
we can also drop the lru cache

drop the data spec into a dict, leave off the lru cache, and actually, if we use _next_scopes[scope], then we get a lookuperror by default as well (if we use it internally, then we dont need the extra text in the value error)

Reply:
Do you think using a mapping is superior? One thing I don't like is that I have to duplicate the names of the scopes. The current implementation doesn't require that.

Can you clarify why do you prefer the mapping?

Reply:
i would use the mapping as its a superior lookup stucture

i would actually declare scopes/next_scopes in terms of dictionary expressions

aka ```py
SCOPE_INDEXES ={s:i for ....}
NEXT_SCOPES = {s: s2 for s, s2 in zip(...), ...}
```